url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/utilities/Generic_Execution.java

```diff
@@ -0,0 +1,165 @@
+                Generic_Execution.waitSychronized(obj, delay);
+                //} catch (InterruptedException e) {
+                //    e.printStackTrace();
+                //}
+            }
+        }
+        System.out.println("Jobs done.");
+        executorService.shutdown();
+
+        try {
+            boolean allterminated = executorService.awaitTermination(maxWait, TimeUnit.MINUTES);
+            System.out.println("All output terminated " + allterminated);
+        } catch (InterruptedException ex) {
+            Logger.getLogger(Generic_Execution.class.getName()).log(Level.SEVERE, null, ex);
+        }
+
+        List<Runnable> unfinishedJobs = executorService.shutdownNow();
+        if (unfinishedJobs.size() > 0) {
+            System.out.println("There are " + unfinishedJobs.size() + " unfinshed jobs");
+        } else {
+            System.out.println("All jobs finished");
+        }
+    }
+
+    /**
+     * 
+     * @param executorService
+     * @param future
+     * @param obj 
+     */
+    public static void shutdownExecutorService(
+            ExecutorService executorService,
+            Future future,
+            Object obj) {
+        long delay = 1000;
+        long maxWait = 10;
+        shutdownExecutorService(
+                executorService,
+                future,
+                obj,
+                delay,
+                maxWait);
+    }
+
+    public static void shutdownExecutorService(
+            ExecutorService executorService,
+            Future future,
+            Object obj,
+            long delay,
+            long maxWait) {
+        long counter = 0;
+        while (!future.isDone()) {
+            System.out.println("Job not done waiting " + delay + " milliseconds having already waited " + counter + " milliseconds");
+            counter += delay;
+            //try {
+            Generic_Execution.waitSychronized(obj, delay);
+            //} catch (InterruptedException e) {
+            //    e.printStackTrace();
+            //}
+        }
+        System.out.println("Job done.");
+        executorService.shutdown();
+        try {
+            boolean allterminated = executorService.awaitTermination(maxWait, TimeUnit.MINUTES);
+            System.out.println("All output terminated " + allterminated);
+        } catch (InterruptedException ex) {
+            Logger.getLogger(Generic_Execution.class.getName()).log(Level.SEVERE, null, ex);
+        }
+        List<Runnable> unfinishedJobs = executorService.shutdownNow();
+        if (unfinishedJobs.size() > 0) {
+            System.err.println("There were " + unfinishedJobs.size() + " unfinshed jobs");
+        } else {
+            System.out.println("All jobs finished");
+        }
+    }
+
+    public static void waitSychronized(Object obj, long timeInMilliseconds) {
+        try {
+            synchronized (obj) {
+                obj.wait(timeInMilliseconds);
+            }
+        } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)